[[ Constants ]] Refactor constants and add infinity constant

livecodeali opened the pull request:
This patch cleans up the implementation of constant lookup, replacing the
string and number value fields in the constant table's Cvalue struct with
a typed union. This enables existing string or number-based constants to
be specified as before, but constants with values that are MCValueRefs to
be represented in the table using specific values of the CvalueType enum.
The lookup mechanism is thereby unaffected. It also removes the MCConstant
syntax class, as it is no longer needed, and uses MCLiteral instead.

The constant values are interned so that there is only ever one instance
of the value in memory.

Previously the `true` and `false` constants were evaluated as strings. This
patch changes them to be evaluated as booleans. The only material difference
this should make is in use of the `is strictly` operator and when the constants
are elements of arrays passed to LCB handlers; in the latter case it is already
the LCB code's responsibility to handle variable typed in incoming array elements.

Furthermore the constant `infinity` has been added, evaluating to +inf (as a
real).

montegoulding commented on extensions/widgets/paletteactions/paletteactions.lcb:
```diff
@@ -700,8 +700,15 @@ private handler drawActions()
 
         variable tMode as String
+        variable tEnabled as optional Boolean
+        // tAction["enabled"] could be a string or boolean
```
@livecodeali @runrevmark this all seems like too much work. It seems to me it would be nice if we had the equivalent of `parse any as <type>` which followed the rules used by LCS for `TryToEval`

montegoulding commented:
@livecode-vulcan review ok 7238b36

livecode-vulcan commented:
💙 review by @montegoulding ok 7238b367ced600030d23a615283257841c28ad68

livecode-vulcan commented:
😎 test success 7238b367ced600030d23a615283257841c28ad68

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/217)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/209)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/209)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/209)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/209)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1731)
* try-community-universal-ios-iphoneos12.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.0/builds/113)
* try-community-universal-ios-iphonesimulator12.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.0/builds/113)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1762)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1754)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1751)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1787)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1728)

runrevmark commented on tests/lcs/core/logic/isstrictly.livecodescript:
```diff
@@ -62,7 +64,8 @@ on TestStrictlyString
    __TestIsStrictly 1, "a string"
    __TestIsStrictly 1.0, "a string"
    __TestIsStrictly infinity, "a string"
```
This patch makes infinity no longer a string but strictly a number

runrevmark commented:
I'm wondering if we should get rid of the MCConstant class entirely - it doesn't really serve any useful purpose anymore.

Instead, it would be better to make constants evaluate as MCLiteral and thus use valuerefs for all constants.

Integers should use MCNumberCreateWithInteger, reals with MCNumberCreateWithReal and strings created with MCNameCreateWithNativeChars (infinity is a real in this case, and null is a string etc.).

Then the constructed value should be passed through MCValueInterAndRelease (which means there will only ever be one instance of the constant in memory at a time) and *then* passed to the MCLiteral constructor.

runrevmark commented:
This will need docs / release notes about the adding of infinity :)

livecodeali committed:
[[ Constants ]] Refactor constant lookup and value creation

This patch cleans up the implementation of constant lookup, replacing the
string and number value fields in the constant table's Cvalue struct with
a typed union. This enables existing string or number-based constants to
be specified as before, but constants with values that are MCValueRefs to
be represented in the table using specific values of the CvalueType enum.
The lookup mechanism is thereby unaffected. It also removes the MCConstant
syntax class, as it is no longer needed, and uses MCLiteral instead.

The constant values are interned so that there is only ever one instance
of the value in memory.

Previously the `true` and `false` constants were evaluated as strings. This
patch changes them to be evaluated as booleans. The only material difference
this should make is in use of the `is strictly` operator and when the constants
are elements of arrays passed to LCB handlers; in the latter case it is already
the LCB code's responsibility to handle variable typed in incoming array elements.

Furthermore the constant `infinity` has been added, evaluating to +inf (as a
real).

livecodeali committed:
[[ PaletteActions ]] Fix potential exec error in palette actions

Previously the palette actions widget assumed the incoming "enabled"
key of an element of the action array and "checked" of a menu item
description were strings - this was only the case if either an
explicit string or one of the boolean constants (which are evaluated
as strings) were put into the array - if an expression had been used
(eg true is true) the widget would have thrown an error. This patch
fixes that possibility, which becomes an inevitability when the
boolean constants are changed to be evaluated as booleans.

livecodeali committed:
[[ Constants ]] Update is strictly tests after constants refactor

This patch tweaks the `is strictly <type>` tests to reflect the fact that
constants are strictly of the expected underlying type - numeral word
constants are integers, true and false are boolean and infinity is a real.

livecodeali committed:
[[ Math ]] Change uses of "inf" to infinity

This patch changes the tests of uses of infinity in math operations 
to use `infinity` rather than `"inf"`, thus saving a type conversion
from string to number.

livecodeali committed:
[[ Constants ]] Add dictionary entry for infinity constant

This patch adds a dictionary entry for the newly added infinity constant

livecodeali committed:
[[ Constants ]] Add release notes for constant changes and addition

This patch adds release notes for:
- the addition of the infinity constant
- the change to the underlying values of the boolean constants

runrevmark commented on docs/dictionary/constant/nothing.lcdoc:
```diff
@@ -0,0 +1,19 @@
+Syntax: nothing
+
+Summary:
+The value with no type
```
I'm not sure this is an accurate description. `The (unique) value representing the notion of no value`.

runrevmark commented on docs/dictionary/constant/nothing.lcdoc:
```diff
@@ -0,0 +1,19 @@
+put nothing is strictly nothing -- true
+
+Description:
+The constant <nothing> is the only value with no type, thus
```
The constant <nothing> is a unique value which is used to represent the notion of no value. It converts to empty when manipulated, but can be distinguished from empty using the `<is strictly nothing>` operator.

runrevmark commented on docs/dictionary/constant/nothing.lcdoc:
```diff
@@ -0,0 +1,19 @@
+operator returns true.
+
+References:
+is strictly (operator)
```
Is the operator not `is strictly nothing`? Or do we lump all the is striclty's together?

livecodeali replied:
We lump them all together. It might be better not to...

runrevmark replied:
That's probably something for when we refine the `is strictly type` operators and add `is strictly`.

runrevmark commented on docs/notes/feature-nothing.md:
```diff
@@ -0,0 +1,5 @@
+# Nothing
+The constant `nothing` has been added to the language.
+
+The `nothing` constant resolved to the internal value kMCNull,
```
The `nothing` constant resolves to a unique value (with unique type) which is used to represent the notion of no value.

runrevmark commented on engine/src/scriptpt.cpp:
```diff
@@ -1310,33 +1309,55 @@ Parse_stat MCScriptPoint::lookup(Script_point t, const LT *&dlt)
 }
 
-bool MCScriptPoint::lookupconstantvalue(const char *& r_value)
+bool MCScriptPoint::lookupconstantintable(uint2& r_position)
```
Since this is wholesale update - just use int rather than uint2 through this (same for cond)... It will result in more efficient code on all platforms.

runrevmark commented on engine/src/scriptpt.cpp:
```diff
@@ -1310,33 +1309,55 @@ Parse_stat MCScriptPoint::lookup(Script_point t, const LT *&dlt)
+    return false;
+}
+
+bool MCScriptPoint::lookupconstantvalue(const char *& r_value)
```
Where is lookupconstantvalue used? It seems a bit of an anachronism.

livecodeali replied:
Yes I wondered about removing it... here is the code:
```
                // MW-2014-11-06: [[ Bug 3680 ]] If we are in explicit var mode, and the token
                //   is not a string literal or a number, then it must be a constant in the constant
                //   table that is the same as its token.
                if (MCexplicitvariables && type == ST_ID)
                {
                    // If the unquoted literal is a recognised constant and the constant's value
                    // is identical (case-sensitively) to the value, it is fine to make it a literal.
                    const char *t_value;
                    if (sp . lookupconstantvalue(t_value) &&
                        MCStringIsEqualToCString(sp . gettoken_stringref(), t_value, kMCStringOptionCompareExact))
                        type = ST_LIT;
                }
```

So actually this patch reintroduces bug 3680 for `local sVar = true` or `constant kFalse = false`. To fix those we need to tweak this to convert constant values to string in `lookupconstantvalue`.

livecodeali replied:
I guess I should just make lookupconstantvalue return a stringref at least

runrevmark replied:
Okay so that is slightly irksome - the current behavior of initializers is not ideal...

When explicitVariables is off:
```
local tVar = empty
put tVar
  => "empty" *not* ""
```

This causes a discontinuity if we were to add constant expression initializers:
```
local tVar = empty => "empty"
local tVar = empty & empty => ""
```

I think changing this behavior can be justified as part of introducing constant expressions; but not for just cleaning up how constants work internally.

One way to resolve this for this patch is to change lookupconstantvalue so it looks up the constant, and handles the cases which have been broken by the refactor - i.e. true/false/infinity return "true"/"false"/"infinity" (as they convert as strings to their constant value).

runrevmark commented on engine/src/scriptpt.cpp:
```diff
@@ -1347,47 +1368,59 @@ Parse_stat MCScriptPoint::lookupconstant(MCExpression **dest)
+            break;
+        case kCValueTypeNull:
+            // Create a stringref that contains an explicit nul character
+            /* UNCHECKED */ MCStringCreateWithBytes((const byte_t*)"",
```
MCNameCreateWithNativeChars...

runrevmark commented on tests/lcs/core/logic/isstrictly.livecodescript:
```diff
@@ -33,36 +33,45 @@ private command __TestIsStrictly pValue, pExpected, pExpectResult
+   __TestIsStrictly nothing, "nothing"
+
+   local tVar
+   __TestIsStrictly tVar, "nothing"
```
Is it worth adding a test for unpassed parameters too? (Which I believe should also be `nothing`). e.g.
```
on Foo pA, pB, pC
  test pB is strictly nothing
end Foo

on callFoo
  foo 1, , 2
end callFoo
```
Note: They might not be currently, but probably should be...

runrevmark replied:
We should probably also check what param(N) returns where N > paramCount as that is the same case as an unpassed argument.

livecodeali replied:
Good idea

runrevmark replied:
So having quickly checked the engine source (MCParameter::eval_ctxt; and EngineEvalParam) it looks like unspecified parameters are kMCEmptyString. I don't think it would hurt to change them to kMCNull though - it would bring consistency with uninitialized variables and has no backwards-compatibility issues beyond the existing uses of 'is strictly' (which changing boolean constants has touched on anyway).

livecodeali replied:
CommandArgumentAtIndex should probably be the same

runrevmark commented on docs/dictionary/constant/nothing.lcdoc:
```diff
@@ -0,0 +1,19 @@
+put nothing is strictly nothing -- true
+
+Description:
+The constant <nothing> is a unique value which is used to represent
```
Okay another attempt...

The constant `<nothing>` is a unique value which is used to represent the absence of any other value. It behaves the same as `<empty>` in all contexts except when tested using the `<is strictly> <nothing>` operator which will only be `<true>` for the `<nothing>` constant.

Variables which have no explicit initializer are set to `<nothing>`, as are explicit handler parameters which have no corresponding argument when called.

runrevmark commented:
@ali There is one other place we could use `nothing` - unset array keys... i.e. tVar["foo"] could be made to return `nothing` if "foo" is not among the keys of tVar.

livecodeali commented:
Ah yes, I was thinking about that - but then what would happen if you do `put nothing into tArray["foo"]`? Would we want that to be equivalent to `delete variable tArray["foo"]`?

runrevmark commented:
If you put nothing into tArray["foo"] then you have an array key "foo" with value nothing as currently posed (so you can tell the difference with 'is among the keys of' though).

The reality is that there is not just the notion of no value, but also no *explicitly assigned* value - i.e. undefined and nothing.

To compare with another well-known scripting language...

JavaScript has `undefined` and `null`. The `undefined` singleton type/value is used for any slot which hasn't been explicitly assigned - e.g. uninitialized variables and uninitialized array keys. Then `null` is a distinct singleton type/value which represents an explicitly set `no value`. The `null` and `undefined` singletons are not identical, but are equal (e.g. similar to the distinction between the notion of `is strictly` and `is`).

As it stands `nothing` is kind of our `undefined`, but that's not quite right - I wonder if we should ditch adding `nothing` right now, and then revisit and introduce `undefined` and `nothing` with appropriate operators and conversions.

runrevmark commented:
As discussed offline - upshot of the above is to pull out the addition of the `nothing` constant as a WIP PR as it touches on areas outwith 'just' refining how constants are mapped internally and adding infinity.

livecodeali committed:
[[ Constants ]] Add constant tests

This patch adds some tests specifically for some constant values. In
particular it tests that variable assigment with a uql which is a
constant name converts appropriately in various contexts.

livecodeali committed:
[[ Constants ]] Add parser tests for uql constant names

This patch adds parser tests to ensure that a parse error is thrown
when a constant name is used as a uql assigment to a variable or
constant if and only if explicitVariables is true and the constant's
name converts (as its constant type) exactly to its value.

livecodeali committed:
[[ Docs ]] Add to breaking changes doc

This patch adds descriptions of the breaking changes caused
by making true and false constants resolve to boolean values
rather than strings, and the introduction of the infinity 
constant, to the breaking changes document.

livecodeali commented:
@runrevmark OK, finally done I think... with added parser tests!

livecodeali commented:
Or will be once I've change the commit messages

runrevmark commented on engine/src/scriptpt.cpp:
```diff
@@ -1310,33 +1309,80 @@ Parse_stat MCScriptPoint::lookup(Script_point t, const LT *&dlt)
+                                        strlen(constant_table[t_position].string),
+                                        &t_constant_value);
+
+            return MCNameIsEqualTo(gettoken_nameref(), *t_constant_value, kMCCompareExact);
```
There's no need to convert to a name her e- just use MCStringIsEqualToCString with gettoken_stringref() and the string from the table.

runrevmark commented on engine/src/scriptpt.cpp:
```diff
@@ -1310,33 +1309,80 @@ Parse_stat MCScriptPoint::lookup(Script_point t, const LT *&dlt)
+                                        strlen(constant_table[t_position].token),
+                                        &t_constant_token);
+
+            return MCNameIsEqualTo(gettoken_nameref(), *t_constant_token, kMCCompareExact);
```
Same here

runrevmark commented:
@livecode-vulcan review ok d84f41e

livecode-vulcan commented:
💙 review by @runrevmark ok d84f41ea7d2b66ccfccf7b2ab66151706e24a78c

livecode-vulcan commented:
😞 test failure d84f41ea7d2b66ccfccf7b2ab66151706e24a78c

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/495)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/487)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/487)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/487)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/487)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2009)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/237)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/236)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2041)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2031)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2029)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2064)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2001)

runrevmark commented:
@livecodeali : So VS doesn't support C99 'named' initializers - i.e. the .foo syntax (that part of C99 is being introduced to C++ in C++20 - and VS doesn't fully support C99). This can be done with a bit of C++ and constexprs though...

Something along the lines of:
```
struct Cvalue
{
    /* Each constructor is labelled constexpr so that it should be
     * evaluated at compile time. Overloading is used to infer the
     * type to use, based on the second argument. */  

    constexpr Cvalue(const char *p_token, integer_t p_integer)
      : token(p_token),
        type(kCValueTypeInteger),
        integer(p_integer)
    {
    }

    constexpr Cvalue(const char *p_token, double p_real)
      : token(p_token),
        type(kCValueTypeReal),
        real(p_real)
    {
    }

    constexpr Cvalue(const char *p_token, const char *p_string)
      : token(p_token),
        type(kCValueTypeString),
        string(p_string)
    {
    }

    /* As CvalueType is an explicit enum type distinct from the other
     * value types, we can use that to construct a Cvalue with explicit
     * type. Note: As this is a constexpr constructor *all* fields must be
     * initialized, we choose integer = 0 for the union in this case. */

    constexpr Cvalue(const char *p_token, CvalueType p_type)
      : token(p_token),
        type(p_type),
        integer(0)
    {
    }

    const char *token;
    CvalueType type;
    union
    {
        double real;
        integer_t integer;
        const char *string;
    };
};

/* The constant_table must be constexpr to force the compiler to generate
 * it at compile time (the rule is that constexpr expressions may be evaluated
 * at compile-time or runtime *unless* they are used as a subexpression of a
 * constexpr in which case they must be evaluated at compile-time - MSVC
 * needs constexpr here to force compile-time eval; other compilers are happy
 * with just const - go figure!). */
constexpr Cvalue constant_table[] =
{
  /* Initializer lists are used as (in C++11) they map to constructors. */
  { "arrow", 29 },
  { "pi", 3.14159 },
  { "backslash", "\\" },
  { "false", kCValueTypeFalse },
};
```

livecodeali commented:
Pretty much exactly what I was going to do until I discovered the `{.foo=x}` syntax :-D

livecodeali commented:
Given my only change was the suggested code, I'm going to self-review to expedite the process
@livecode-vulcan review ok c7e928f

livecode-vulcan commented:
💙 review by @livecodeali ok c7e928fd3f073f0c1f62c8b456d0c5bb8f23337d

livecode-vulcan commented:
😞 test failure c7e928fd3f073f0c1f62c8b456d0c5bb8f23337d

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/500)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/492)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/492)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/492)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/492)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2014)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/242)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/241)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2046)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2036)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2034)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2069)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2006)

runrevmark commented:
Okay so this is still failing :( I *think* it needs to be `constexpr const Cvalue constant_values[]` so the storage class in the definition matches the external declarations (since it succeeded on the other compilers I guess they treat constexpr as implicit const - which MSVC clearly does not!).

livecodeali commented:
@runrevmark Ok, have made that change!

runrevmark commented:
Okay... Let's try this again :)

runrevmark commented:
@livecode-vulcan review ok a217659

livecode-vulcan commented:
💙 review by @runrevmark ok a217659c7ca86001d49c8779047657692abd3279

livecode-vulcan commented:
😞 test failure a217659c7ca86001d49c8779047657692abd3279

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/502)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/494)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/494)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/494)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/494)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2016)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/244)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/243)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2048)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2038)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2036)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2071)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2008)

runrevmark commented:
Right so - MCVC does not like mixing constexpr and const - whatever mangling extern constexpr const Cvalue... does to the name is incompatible with trying to do an 'extern const Cvalue ...' from elsewhere (and you can't have constexpr without a definition). So the following appears to work...

Change all `extern const Cvalue constant_table[];` in various places to `
extern const Cvalue *constant_table;`.

Change the constexpr to - `static constexpr const Cvalue constant_table_values[] =` and remove the extern in the same file.

Then do this:
```
extern const Cvalue *constant_table = constant_table_values;
extern const uint4 constant_table_size = ELEMENTS(constant_table_values);
```

This works because we are keeping the constexpr static/local to the translation unit but defined - and then creating a extern Cvalue* to it - which *does* work across translation units. (Remember that foo[] is actually a foo*).

runrevmark commented:
Okay so having tested 'locally' the above changes do seem to now compile with MSVC :)

runrevmark commented:
@livecode-vulcan review ok 76ad61e

livecode-vulcan commented:
💙 review by @runrevmark ok 76ad61efea4e7a7baa3fd55a2a812d90b0c2a4dc

livecode-vulcan commented:
😎 test success 76ad61efea4e7a7baa3fd55a2a812d90b0c2a4dc

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/506)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/498)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/498)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/498)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/498)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2020)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/248)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/247)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2052)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2042)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2040)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2075)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2012)

livecodeali commented:
Hurrah!
